pkg/provider/consulcatalog/config.go

```diff
@@ -120,8 +121,9 @@ func (c *connectCert) getLeaf() tls.Certificate {
 
 func (c *connectCert) serverTransport(sname string) *dynamic.ServersTransport {
 	return &dynamic.ServersTransport{
-		ServerName: sname,
-		RootCAs:    c.getRoot(),
+		ServerName:         sname,
+		InsecureSkipVerify: c.insecureSkipVerify,
+		RootCAs:            c.getRoot(),
```

Comment:
As a follow up to the previous config. Shouldn't configuring the roots here allow traefik to trust those?

Reply:
> As a follow up to the previous comment. Shouldn't configuring the roots here allow traefik to trust those?

I had problems getting it to work, but I just realised that it's probably due to this:
```Go
//roundtripper.go 142-149
if cfg.InsecureSkipVerify || len(cfg.RootCAs) > 0 || len(cfg.ServerName) > 0 || len(cfg.Certificates) > 0 {
	transport.TLSClientConfig = &tls.Config{
		ServerName:         cfg.ServerName,
		InsecureSkipVerify: cfg.InsecureSkipVerify,
		RootCAs:            createRootCACertPool(cfg.RootCAs),
		Certificates:       cfg.Certificates.GetCertificates(),
	}
}
```
So TLS is only established if `InsecureSkipVerify` is set to `true`

Reply:
But shouldn't `RootCAs` also be larger than zero?

Reply:
> But shouldn't `RootCAs` also be larger than zero?

I'm a bit confused about what you mean. Could you elaborate? :)

Reply:
`RootCAs` provides the CAs used to verify the connect service certs. The whole point of connect is to have a secure TLS connection; but since Consul connect has a non-trusted autogenerated CA one needs to fetch the RootCAs of the Connect CA and set those on the transport. This code https://github.com/traefik/traefik/pull/7407/files#diff-c3e25341e36d91fba700c36777fee7b575bb7004a9521a76ec17d2fb3f748364R141-R143 should publish `connectCert.root` which should be non-empty and passed to the servers transport.

Setting `InsecureSkipVerify` is completely against the idea of connect :D

Reply:
> `RootCAs` provides the CAs used to verify the connect service certs. The whole point of connect is to have a secure TLS connection; but since Consul connect has a non-trusted autogenerated CA one needs to fetch the RootCAs of the Connect CA and set those on the transport. This code https://github.com/traefik/traefik/pull/7407/files#diff-c3e25341e36d91fba700c36777fee7b575bb7004a9521a76ec17d2fb3f748364R141-R143 should publish `connectCert.root` which should be non-empty and passed to the servers transport.
> 
> Setting `InsecureSkipVerify` is completely against the idea of connect :D

Thanks for the explanation. I'm not very strong in TLS, so I appreciate that you help me out here :smiley: 

I think the problem is, that Traefik uses the ip of the service, rather than the Consul DNS name. The certificates are only valid for `*.consul` domains

To fix this, Traefik will need to be aware of the Consul DNS and use the internal domain name to resolve the service

Reply:
I think I have a way to fix this, but might not have time to do it today

Reply:
> To fix this, Traefik will need to be aware of the Consul DNS and use the internal domain name to resolve the service

Yes you probably should set `ServerName` on the transport? This will fix the SNI issues (ie inside the TLS handshake)

Reply:
> I think I have a way to fix this, but might not have time to do it today

No worries, I'll try to get this running myself today as well to provide better feedback. Thank you for working on it!

Reply:
> > To fix this, Traefik will need to be aware of the Consul DNS and use the internal domain name to resolve the service
> 
> Yes you probably should set `ServerName` on the transport? This will fix the SNI issues (ie inside the TLS handshake)

That's way smarter, than what I was thinking. I'll give it a go :+1:

Reply:
I looked through the source for the Consul package an it seems like Hashicorp themselves do their own validation

```Go
// We have to set this since otherwise Go will attempt to verify DNS names
// match DNS SAN/CN which we don't want. We hook up VerifyPeerCertificate to
// do our own path validation as well as Connect AuthZ.
InsecureSkipVerify: true,
```

Source: https://github.com/hashicorp/consul/blob/91ee7990ccf27ab7fccff746f8df0df65c3c583d/connect/tls.go#L75-L78

Reply:
Looked through the source for the Consul package and it seems like Hashicorp also skip verification, instead implementing their own verification mechanism:

```Go
// We have to set this since otherwise Go will attempt to verify DNS names
// match DNS SAN/CN which we don't want. We hook up VerifyPeerCertificate to
// do our own path validation as well as Connect AuthZ.
InsecureSkipVerify: true,
```

[Source](https://github.com/hashicorp/consul/blob/91ee7990ccf27ab7fccff746f8df0df65c3c583d/connect/tls.go#L75-L78)

Reply:
Yes but they implement their own validation apparently. I really hope that this is doable with `ServerName` because otherwise it will get quite some fun :)

Reply:
> Yes but they implement their own validation apparently. I really hope that this is doable with `ServerName` because otherwise it will get quite some fun :)

I don't think there's an easy fix with the ServerName, since the validation uses the globally defined DNS to do validation. It might be possible to do by changing the DNS in the `net.Dialer` configuration, but I think the work required for that solution makes it more preferable to just pass the `DialTLS` to the transporter as @Gufran also suggested [here](https://github.com/traefik/traefik/pull/7407#issuecomment-726007829)

It's a bigger change. The quick fix is to simply add one or more fields to the `ServersTransport` struct that configures which `Dial` function is used, but I think that's inelegant, since it's only one provider, that would need it

I think that it's best to merge the PR as it is right now and then continue the discussion in the upstream PR. Hopefully we can get some input from the maintainers there. I'll sign up for the next scheduled "office hours" if that's needed :)

Reply:
I tried your branch now and I get:
```
ERRO[2021-01-19T15:17:24+01:00] Failed to register traefik as Connect Native service in consul catalog. failed to register service in consul catalog. Unexpected response code: 405 (method GET not allowed)  providerName=consulcatalog
```
the consul logs show:
```
2021-01-19T14:17:19.230Z [ERROR] agent.http: Request error: method=GET url=/v1/agent/service/register from=172.22.3.2:43978 error="method GET not allowed"
```
any idea why it would try to register via GET?

I have set the following config:
```
TRAEFIK_PROVIDERS_CONSULCATALOG_ENDPOINT_ADDRESS=consul.domain
TRAEFIK_PROVIDERS_CONSULCATALOG_ENDPOINT_TOKEN=my_token
TRAEFIK_PROVIDERS_CONSULCATALOG_EXPOSEDBYDEFAULT=false
TRAEFIK_PROVIDERS_CONSULCATALOG_CONNECT_CONNECTAWARE=true
TRAEFIK_PROVIDERS_CONSULCATALOG_SERVICENAME=traefik-test
```

Reply:
Oh that was an issue on my side, needed to set a proper scheme for the connection.

Reply:
> I think that it's best to merge the PR as it is right now and then continue the discussion in the upstream PR. 

Jupp that sound like I good plan

Reply:
Ah few things; would it make to apply the following patch:
```
diff --git a/pkg/provider/consulcatalog/config.go b/pkg/provider/consulcatalog/config.go
index 6df6803c..62345d6c 100644
--- a/pkg/provider/consulcatalog/config.go
+++ b/pkg/provider/consulcatalog/config.go
@@ -4,6 +4,8 @@ import (
        "context"
        "errors"
        "fmt"
+       "net"
+
        "github.com/hashicorp/consul/api"
        "github.com/traefik/traefik/v2/pkg/config/dynamic"
        "github.com/traefik/traefik/v2/pkg/config/label"
@@ -11,7 +13,6 @@ import (
        "github.com/traefik/traefik/v2/pkg/provider"
        "github.com/traefik/traefik/v2/pkg/provider/constraints"
        "github.com/traefik/traefik/v2/pkg/tls"
-       "net"
 )
 
 func (p *Provider) buildConfiguration(ctx context.Context, items []itemData, certInfo *connectCert) *dynamic.Configuration {
@@ -67,7 +68,7 @@ func (p *Provider) buildConfiguration(ctx context.Context, items []itemData, cer
                        confFromLabel.HTTP.ServersTransports = make(map[string]*dynamic.ServersTransport)
                }
 
-               if certInfo != nil {
+               if certInfo != nil && item.ExtraConf.ConnectEnabled {
                        confFromLabel.HTTP.ServersTransports[connectTransportName(item.Name)] = certInfo.serverTransport(connectTransportName(item.Name))
                }
```

This way we'd only recreate extra server transports for connect enabled services (and not for every service).

Also as soon as the catalog configuration is refreshed I get:
```
DEBU[2021-01-19T15:52:27+01:00] '500 Internal Server Error' caused by: x509: certificate is not valid for any names, but wanted to match connect-tls-hello-sidecar-proxy 
```
on a service which did work before; so there seem to be some issues on the way.